lib/datasource.js

```diff
@@ -538,24 +525,112 @@ DataSource.prototype.defineRelations = function(modelClass, relations) {
       if ((targetModel && !isModelDataSourceAttached(targetModel)) ||
           (throughModel && !isModelDataSourceAttached(throughModel))) {
         // Create a listener to defer the relation set up
-        createListener(rn, r, targetModel, throughModel);
+        deferRelationSetup(relationName, r, targetModel, throughModel);
       } else {
         // The target model is resolved
         var params = traverse(r).clone();
-        params.as = rn;
-        params.model = polymorphicName || targetModel;
+        params.as = relationName;
+        params.model = targetModel;
         if (throughModel) {
           params.through = throughModel;
         }
-        modelClass[r.type].call(modelClass, rn, params);
+        modelClass[r.type].call(modelClass, relationName, params);
       }
     });
   }
 };
 
+function validateRelation(relationName, relation) {
+  var rn = relationName;
+  var r = relation;
+  var code;
+
+  // allows for method injection via relation definition in unit tests
+  getRelationValidationMsg = r.getRelationValidationMsg || getRelationValidationMsg;
+
```

Comment:
i think you meant 
> the real `getRelationValidationMsg` function is not tested.

Hmm I see your concern. that would only affect the error msg though, a validation error would still be thrown. If we want it to be less human error prone, and keep not checking the exact message names, then i'll refactor without `getRelationValidationMsg`.